src/metabase/models/card.clj

```diff
@@ -237,38 +248,40 @@
-    (check-field-filter-fields-are-from-correct-database changes)
-    ;; Make sure the Collection is in the default Collection namespace (e.g. as opposed to the Snippets Collection namespace)
-    (collection/check-collection-namespace Card (:collection_id changes))
-    (params/assert-valid-parameters changes)
-    (params/assert-valid-parameter-mappings changes)
-    ;; additional checks (Enterprise Edition only)
-    (@pre-update-check-sandbox-constraints changes)
-    ;; create Action and QueryAction when is_write is set true
-    (create-actions-when-is-writable! changes)
-    ;; delete Action and QueryAction when is_write is set false
-    (delete-actions-when-not-writable! changes)))
+  (let [card (db/select-one Card :id id)]
+    (u/prog1 changes
+      ;; if the Card is archived, then remove it from any Dashboards
+      (when archived?
+        (db/delete! 'DashboardCard :card_id id))
+      ;; if the template tag params for this Card have changed in any way we need to update the FieldValues for
+      ;; On-Demand DB Fields
+      (when (get-in changes [:dataset_query :native])
+        (let [old-param-field-ids (params/card->template-tag-field-ids (select-keys card [:dataset_query]))
+              new-param-field-ids (params/card->template-tag-field-ids changes)]
+          (when (and (seq new-param-field-ids)
+                     (not= old-param-field-ids new-param-field-ids))
+            (let [newly-added-param-field-ids (set/difference new-param-field-ids old-param-field-ids)]
+              (log/info "Referenced Fields in Card params have changed. Was:" old-param-field-ids
+                        "Is Now:" new-param-field-ids
+                        "Newly Added:" newly-added-param-field-ids)
+              ;; Now update the FieldValues for the Fields referenced by this Card.
+              (field-values/update-field-values-for-on-demand-dbs! newly-added-param-field-ids)))))
+      ;; make sure this Card doesn't have circular source query references if we're updating the query
+      (when (:dataset_query changes)
+        (check-for-circular-source-query-references changes))
+      ;; Make sure any native query template tags match the DB in the query.
+      (check-field-filter-fields-are-from-correct-database changes)
+      ;; Make sure the Collection is in the default Collection namespace (e.g. as opposed to the Snippets Collection namespace)
+      (collection/check-collection-namespace Card (:collection_id changes))
+      (params/assert-valid-parameters changes)
+      (params/assert-valid-parameter-mappings changes)
+      ;; additional checks (Enterprise Edition only)
+      (@pre-update-check-sandbox-constraints changes)
+      ;; create Action and QueryAction when is_write is set true
+      (create-actions-when-is-writable! changes)
+      ;; delete Action and QueryAction when is_write is set false
+      (delete-actions-when-not-writable! changes)
+      (assert-valid-model (merge card changes)))))
```

Comment:
I think we could save some DB calls here. 
Wdyt if we only run this assert when the `changes` contains `:template-tags`?

the idea is we try to not fetch the card every time we do an update if it's not a must.

Reply:
You're right, we could save some DB calls. I've fixed that in the [latest commit](https://github.com/metabase/metabase/pull/24902/commits/12d25253b535f7c94221707fc073de8ab2517fad).

I don't really understand the trade-off between adding a DB call vs making the code simpler though. I guess I'll learn that with time?